File: Week 05/id_011/decode-ways.go
package algorithm00401

func numDecodings(s string) int {
	if len(s) == 0 || s[:1] == "0" {
		return 0
	}
	a, b := 1, 1
	for i := 1; i < len(s); i++ {
		if s[i:i+1] == "0" {
			a = 0
		}
		if s[i-1:i] == "1" || (s[i-1:i] == "2" && s[i:i+1] <= "6") {
			a = a + b
			b = a - b
		} else {
			b = a
		}
	}
	return a
}

File: Week 08/id_306/DecodeWays.java
package sf.week8;

/**
 * Created by LynnSun on 2019/12/6.
 * 力扣题目地址：https://leetcode-cn.com/problems/decode-ways
 */
public class DecodeWays {
    public int numDecodings(String s) {
        if (s == null || s.length() == 0) {
            return 0;
        }
        int len = s.length();

        int help = 1;
        int res = 0;
        if (s.charAt(len - 1) != '0') {
            res = 1;
        }
        for (int i = len - 2; i >= 0; i--) {
            if (s.charAt(i) == '0') {
                help = res;
                res = 0;
                continue;
            }
            if ((s.charAt(i) - '0') * 10 + (s.charAt(i + 1) - '0') <= 26) {
                res += help;
                //help用来存储res以前的值
                help = res-help;
            } else {
                help = res;
            }

        }
        return res;
    }
}
